File: moseoh/_4주차/괄호회전하기/java/Solution.java
package _4주차.괄호회전하기.java;

import java.util.regex.Pattern;

class Solution {
    public static void main(String[] args) {
        String s1 = "[](){}";
        assert new Solution().solution(s1) == 3 : "실패";

        String s2 = "}]()[{";
        assert new Solution().solution(s2) == 2 : "실패";

        String s3 = "[)(]";
        assert new Solution().solution(s3) == 0 : "실패";

        String s4 = "}}}";
        assert new Solution().solution(s4) == 0 : "실패";
    }

    public int solution(String s) {
        int answer = 0;

        for (int i = 0; i < s.length(); i++) {
            s = s.substring(1) + s.charAt(0);
            if (isRight(s)) answer++;
        }
        return answer;
    }

    private boolean isRight(String str) {
        String regex = "(\\(\\)|(\\{\\})|(\\[\\]))";
        Pattern pattern = Pattern.compile(regex);
        while (pattern.matcher(str).find()) {
            str = str.replaceAll(regex, "");
        }
        return str.length() == 0;
    }
}

File: moseoh/_4주차/괄호회전하기/kotlin/Solution.kt
package _4주차.괄호회전하기.kotlin

fun main() {
    val s1 = "[](){}"
    check(Solution().solution(s1) == 3) { "오답" }

    val s2 = "}]()[{"
    check(Solution().solution(s2) == 2) { "오답" }

    val s3 = "[)(]"
    check(Solution().solution(s3) == 0) { "오답" }

    val s4 = "}}}"
    check(Solution().solution(s4) == 0) { "오답" }
}

class Solution {
    fun solution(s: String): Int {
        var answer = 0
        var str = s

        for (i in s.indices) {
            str = str.slice(1 until str.length) + str[0]
            if (isRight(str)) {
                answer++
            }
        }
        return answer
    }

    private fun isRight(str: String): Boolean {
        var s = str
        val regex = Regex("(\\(\\))|(\\{\\})|(\\[\\])")
        while (s.contains(regex)) {
            s = s.replace(regex, "")
        }
        return s.isEmpty()
    }
}
